backfill/2024-03-23-clients_last_seen/telemetry_derived_clients_last_seen_v2_20240322/backfill_clients_last_seen_v2.py

```diff
@@ -0,0 +1,330 @@
+        udf.coalesce_adjacent_days_28_bits(
+          _previous.days_created_profile_bits,
+          _current.days_created_profile_bits
+        ) AS days_created_profile_bits,
+        udf.combine_experiment_days(
+          _previous.days_seen_in_experiment,
+          _current.days_seen_in_experiment
+        ) AS days_seen_in_experiment
+      )
+    FROM
+      _current
+    FULL JOIN
+      _previous
+      USING (client_id)
+    LEFT JOIN
+      `moz-fx-data-shared-prod.telemetry_derived.clients_first_seen_v2` AS cfs
+      USING (client_id)
+"""
+
+CREATE_TABLE_QUERY = """
+    CREATE OR REPLACE TABLE
+      `{project_id}.{dataset}.{table}_{sample_id}`
+    PARTITION BY
+        submission_date
+    CLUSTER BY
+        normalized_channel,
+        sample_id
+    AS
+    SELECT
+        *
+    FROM
+        `{project_id}.{dataset}.{table}`
+    WHERE
+        FALSE;
+"""
+
+parser = ArgumentParser()
+parser.add_argument(
+    "--project_id",
+    "--project-id",
+    default="moz-fx-data-shared-prod",
+    help="ID of the project in which to find tables",
+)
+parser.add_argument(
+    "--dataset",
+    default="backfills_staging_derived",
+    help="Dataset name to create clients_first_seen_table in",
+)
+parser.add_argument(
+    "--table",
+    default="telemetry_derived_clients_last_seen_v2_20230322",
+    help="Name of the destination table to be created",
+)
+parser.add_argument(
+    "--parallelism",
+    "-p",
+    default=10,
+    help="Number of threads run in parallel",
+)
+parser.add_argument(
+  "--dry_run",
+  default=False,
+  help="Issue a dry run"
+)
+parser.add_argument(
+    "--start_date",
+    "--start-date",
+    "-s",
+    help="First date to be backfilled",
+    type=click.DateTime(formats=["%Y-%m-%d"]),
+    default='2023-03-23'
+)
+parser.add_argument(
+    "--end_date",
+    "--end-date",
+    "-e",
+    help="Last date to be backfilled",
+    type=click.DateTime(formats=["%Y-%m-%d"]),
+    # default=str(date.today()),
+    default='2023-03-23',
+)
+
+
+def _backfill_staging_table(client, job_config, project_id, dataset, table, bigquery_schema, submission_date, sample_id):
+    """Backfill for a submission_date, sample_id combination."""
+    full_table_id = f"{project_id}.{dataset}.{table}_{sample_id}"
+    try:
+        table = client.get_table(full_table_id)
+    except NotFound:
+        table = bigquery.Table(full_table_id)
+        table.schema = bigquery_schema
+
+    if not table.created:
+        client.create_table(table)
+        click.echo(f"Destination table {full_table_id} created.")
+
+    print(f"Running a backfill for sample_id={sample_id} to {full_table_id}")
+    print(PARTITION_QUERY.format(submission_date=submission_date, sample_id=sample_id, full_table_id=full_table_id))
+    client.query(
+        PARTITION_QUERY.format(submission_date=submission_date, sample_id=sample_id, full_table_id=full_table_id),
+        job_config=job_config
+    ).result()
```

Comment:
I assume once these tables per sample_id have been verified, they'll be copied manually into a single table?

Reply:
Yes, that's the idea.